src/allmydata/storage/http_client.py

```diff
@@ -114,6 +115,33 @@ def __init__(self, code, *additional_args):
 }
 
 
+@define
+class _LengthLimitedCollector:
+    """
+    Collect data using ``treq.collect()``, with limited length.
+    """
+
+    remaining_length: int
+    f: BytesIO = field(factory=BytesIO)
+
+    def __call__(self, data: bytes):
+        if len(data) > self.remaining_length:
+            raise ValueError("Response length was too long")
+        self.f.write(data)
+
+
+def limited_content(response, max_length: int = 30 * 1024 * 1024) -> Deferred:
+    """
+    Like ``treq.content()``, but limit data read from the response to a set
+    length.  If the response is longer than the max allowed length, the result
+    fails with a ``ValueError``.
+    """
+    collector = _LengthLimitedCollector(max_length)
```

Comment:
For the second point, if an exception is raised it... might end up in a Deferred (which I think is what I have tests for), or in some cases it might end up raising an exception directly from `content_length`. It depends! So, not so good... And that doesn't even answer the question... So clearly need a bit more work here.

Reply:
I don't think there's any way to tell a response to close... so I think we're stuck with garbage collection.

Reply:
OK, I actually figured out how to close the connection. So I filed https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3908#ticket for the first point, and will go deal with the second point now.

Reply:
Looks like this has to happen inside `treq`; I filed https://github.com/twisted/treq/issues/347, and will go implement PR for it, but for purposes of this PR we're done.

Reply:
OK, PR submitted. If it gets merged, this will be handled automatically by treq; if status quo continues, not much we can do without a lot more work.